### DISCUSSION TITLE
Set Primary Site dynamically (via .env or CLI)

### DISCUSSION THREAD

**Discussion by mandrasch** [craftcms/cms#14135] [Ideas] [answered]
Hi! 

We have a new project where we would love to use the following deployment setting:

1. One Craft CMS project repository
2. Add multiple sites to it with different languages
3. Deploy it to multiple server instances which have their own domains
4. Set the Primary Site dynamically via .env (or CLI command on deployment)

**Technical Draft**

<img width="890" alt="image" src="https://github.com/craftcms/cms/assets/777278/668e7bde-8e16-4dc5-9437-8612b054caeb">

**Already Possible**

✅ The Status for a site can be dynamically defined via `.env`

<img width="391" alt="image" src="https://github.com/craftcms/cms/assets/777278/675cdf4c-e70a-4f3c-b288-6894cccd6058">

✅ As well as the Base URL, dynamic settings via alias in `general.php`:

<img width="296" alt="image" src="https://github.com/craftcms/cms/assets/777278/2fe675a2-526c-4ac3-aa94-d9b614441395">

```php
 '@web-it' => App::env('SITE_IT_URL'),
```

**What's missing**

But to achieve our goal, the setting `Make this the primary site` should also be dynamically switchable (because primary sites can't be disabled)

<img width="427" alt="image" src="https://github.com/craftcms/cms/assets/777278/c6c1d2c6-c4a6-40a0-a758-13ebbeb79e83"> <br>

<img width="388" alt="image" src="https://github.com/craftcms/cms/assets/777278/ecb39b87-9567-4360-b977-d7cda5a76b88">

**Question** 

Is this somehow achievable / could this be a new CraftCMS feature?  

(We already thought about moving the primary english site to a sub-path like italian-shop.it/en/, but disabling it completely would be much more easier for our use case and our SEO requirements.)

Thanks very much in advance for information and looking into this! 👍 
Best regards, Matthias

**Comment by brandonkelly** [comment, marked as answer]
Unfortunately this isn’t possible. All of your non-localized content (users, etc.) get stored to the primary site, and need to be reassigned whenever the primary site changes. So it’s not something that could be easily toggled via an environment variable.

**Reply by mandrasch** [reply]
Hi @brandonkelly, 
thanks very much for the quick response, much appreciated!

A follow-up question:

Is there a possibility to set the language dynamically for the (primary) site instead?

Thanks!

**Reply by brandonkelly** [reply]
That should be more doable.

**Reply by mandrasch** [reply]
Hi @brandonkelly!

oh wow, nice! 🤩 👍 

We talked about it and would be very interested in this setting for our upcoming project. 

This would definitely solve our upcoming project challenge where we just need to flexible define Sites, their URL and their language via .env. 

Could you give me a hint on what I could do - should I submit another Feature Request or something else?

Thanks very much in advance!
Best regards, Matthias

**Reply by brandonkelly** [reply]
I spec’d out the feature yesterday and it’s a larger change than we should be making in Craft 4.x, so I’ve added it to the Craft 5 list. We should be able to get it in there in time for the beta, soon.

**Reply by mandrasch** [reply]
Hi @brandonkelly, thanks so much for looking into this and adding it to the v5 list, awesome! 👍 👍 🎉 

In our current project (with commerce) we won't be really able to wait for v5 though.

Is there another recommended solution for a flexible deployment scenario like ours with v4?

Our goal is to have 

- one codebase (one repository)
- deploy it to multiple targets (where the sites and the languages are configured via .env / dynamically via PHP)

We're thankful for any hints! 

_(I already researched the web and discord, also watched an older episode of craft quest with 
"Question: Share codebase between different sites" https://craftquest.io/livestreams/craftquest-on-call-15-questions-answered - but haven't really found something like this)_

**Comment by brandonkelly** [comment]
Craft 5.0.0-alpha.10 has been released with support for setting sites’ languages to environment variables (via #14235).
